spec/services/pii/re_encryptor_spec.rb

```diff
@@ -1,33 +1,70 @@
 require 'rails_helper'
 
 RSpec.describe Pii::ReEncryptor do
+  let(:expected_pii_attributes) do
+    pii_attributes = Pii::Attributes.new
+    pii_attributes[:ssn] = ssn
+    pii_attributes
+  end
+
+  let(:pii) { { ssn: ssn } }
+  let(:pii_cacher) { Pii::Cacher.new(user, user_session) }
+  let(:ssn) { '1234' }
+  let(:user) { profile.user }
+  let(:user_session) { {}.with_indifferent_access }
+
+  before do
+    allow(Pii::Cacher).to receive(:new).and_return(pii_cacher)
+    allow(pii_cacher).to receive(:fetch).and_call_original
+
+    if user.active_profile
+      allow(user.active_profile).to receive(:encrypt_recovery_pii).and_call_original
+      allow(user.active_profile).to receive(:save!).and_call_original
+    end
+
+    if user.pending_profile
+      allow(user.pending_profile).to receive(:encrypt_recovery_pii).and_call_original
+      allow(user.pending_profile).to receive(:save!).and_call_original
+    end
+
+    Pii::ProfileCacher.new(user, user_session).save_decrypted_pii(pii, profile.id)
+  end
+
   describe '#perform' do
-    it 're-encrypts PII using new code' do
-      profile = create(:profile, :active, :verified, pii: { ssn: '1234' })
-      user = profile.user
-      user_session = {
-        decrypted_pii: { ssn: '1234' }.to_json,
-      }
+    context 'when the user has an active profile' do
+      let(:profile) { create(:profile, :active, :verified, pii: pii) }
 
-      pii_attributes = Pii::Attributes.new
-      pii_attributes[:ssn] = '1234'
+      before do
+        Pii::ReEncryptor.new(user: user, user_session: user_session).perform
+      end
 
-      expect(user.active_profile).to receive(:encrypt_recovery_pii).
-        with(pii_attributes).and_call_original
-      expect(user.active_profile).to receive(:save!).and_call_original
+      it 'fetches the PII from the correct profile' do
+        expect(pii_cacher).to have_received(:fetch).with(profile.id)
+      end
 
-      Pii::ReEncryptor.new(user: user, user_session: user_session).perform
+      it 're-encrypts PII using new code' do
```

Comment:
I wonder if this could be done with a little less mocking:

1. create a user/profile 
2. call perform
3. verify you can recover the pii using the newly generated personal key

The only gotcha here is you need to use different PII to distinguish pending from active profiles.

Reply:
Yup. Paired w/ Sonia and got advice from Hooper to do that.